accounting/accounting_cli_functions.py

```diff
@@ -248,6 +248,196 @@ def view_jobs_before_end_time(conn, time_before, output_file):
+                pass
+            # add up all of the shares allocated to this parent's sub banks
+            else:
+                for index, row in dataframe.iterrows():
+                    total_sub_bank_shares += row["shares"]
+            # if the total amount of shares in all the sub banks is
+            # equal to the parent bank's shares, it has already reached
+            # its max allocation, and thus the bank won't be added
+            if int(total_sub_bank_shares) >= int(parent_bank_shares):
+                print("Total shares for parent account are already maxed")
+                sys.exit(-1)
+            # if the current total amount of shares in all the sub banks
+            # PLUS the number of shares trying to be added is greater than the
+            # parent's banks shares, the parent account WILL exceed its max
+            # allocation and thus the account won't be added
+            if int(total_sub_bank_shares) + int(shares) > int(parent_bank_shares):
+                print("Total shares will exceed parent account's allocation")
+                sys.exit(-1)
+
+        except pd.io.sql.DatabaseError as e_database_error:
+            print(e_database_error)
+
+    # insert the bank values into the database
+    try:
+        conn.execute(
+            """
+            INSERT INTO bank_table (
+                bank,
+                parent_bank,
+                shares
+            )
+            VALUES (?, ?, ?)
+            """,
+            (bank, parent_bank, shares),
+        )
+        # commit changes
+        conn.commit()
+    # make sure entry is unique
+    except sqlite3.IntegrityError as integrity_error:
+        print(integrity_error)
+
+
+def view_bank(conn, bank):
+    try:
+        # get the information pertaining to a bank in the Accounting DB
+        select_stmt = "SELECT * FROM bank_table where bank=?"
+        dataframe = pd.read_sql_query(select_stmt, conn, params=(bank,))
+        # if the length of dataframe is 0, that means
+        # the bank specified was not found in the table
+        if len(dataframe.index) == 0:
+            print("Bank not found in bank_table")
+        else:
+            print(dataframe)
+    except pd.io.sql.DatabaseError as e_database_error:
+        print(e_database_error)
+
+
+def delete_bank(conn, bank):
+    last_parent_bank_seen = bank
+
+    # delete bank from bank_table
+    delete_stmt = "DELETE FROM bank_table WHERE bank=?"
+    cursor = conn.cursor()
+    cursor.execute(delete_stmt, (bank,))
+
+    # if bank is a parent_bank, find all
+    # sub banks and remove them as well
+    while True:
+        select_stmt = "SELECT bank FROM bank_table WHERE parent_bank=?"
+        dataframe = pd.read_sql_query(
+            select_stmt, conn, params=(last_parent_bank_seen,)
+        )
+        for index, row in dataframe.iterrows():
+            cursor.execute(delete_stmt, (row["bank"],))
+            last_parent_bank_seen = row["bank"]
```

Comment:
If multiple rows are deleted, then some banks will be missed, right?  Take the following bank hierarchy for example:
```
      A
      |
   |-----|
   B     C
   |     |
 |---| |---|
 D   E F   G
```

If `A` is passed into `delete_bank`, then it will be deleted by the first delete execute.  The first `select_stmt` will return `B, C`.  `B` will be deleted and `last_parent_bank_seen` set to `B`.  `C` will be deleted and `last_parent_bank_seen` set to `C`.  The while loop will go to the next iteration, and the select statement will return `F` and `G` (which will then be deleted).   `D` and `E` will still be left in the database.

My recommendation would be to either make this function recursive or to make `last_parent_bank_seen` a queue that parent banks get pushed onto and popped from (rather than just a normal variable).

I also recommend adding a deep bank hierarchy example like the one in this comment to the test suite to ensure there aren't any orphaned banks after a delete.

Reply:
Thanks for explaining this to me @SteVwonder! This is really helpful.

You're absolutely right. The `delete-bank` function will need some rework. Maybe we can take the same stance as the issue with the error propagation - I can create an issue to track this, and for now, the `delete-bank` subcommand will be refactored to just delete one bank at a time, regardless if it is a parent bank or not. Then once this lands, I can work on this in the immediate future. The reason being I think I might need some more time on figuring this out. 

Thoughts? Would that be okay with you?

Reply:
> Thoughts? Would that be okay with you?

Sounds good to me :)